packages/sdk/src/libs/multicall.ts

```diff
@@ -0,0 +1,102 @@
+import { multicall } from "./clients";
+import { Contract } from "ethers";
+import type { SignerOrProvider } from "..";
+import { zip } from "lodash";
+
+export type Call = {
+  method: string;
+  args?: any[];
+};
+
+export type Request = {
+  contractInstance: Contract;
+  call: Call;
+};
+export type EncodedResponse = string;
+export type EncodedRequest = {
+  target: string;
+  callData: string;
+};
+
+interface CoreState {
+  requests: Request[];
+  multicallClient: multicall.Instance;
+}
+
+// Core multicall functionality, not meant for use outside this file
+class Core implements CoreState {
+  public requests: Request[];
+  public multicallClient: multicall.Instance;
+  constructor(state: CoreState) {
+    // make a copy of this so we dont mutate the original
+    this.requests = [...state.requests];
+    this.multicallClient = state.multicallClient;
+  }
+
+  // encode requests to multicall contract
+  private encodeRequest(request: Request) {
+    const { contractInstance, call } = request;
+    return {
+      target: contractInstance.address,
+      callData: contractInstance.interface.encodeFunctionData(call.method, call.args),
+    };
+  }
+
+  // decode response from multicall contract
+  private decodeResponse(request: Request, response: EncodedResponse) {
+    const { contractInstance, call } = request;
+    return contractInstance.interface.decodeFunctionResult(call.method, response);
+  }
+
+  public add(contractInstance: Contract, call: Call) {
+    this.requests.push({ contractInstance, call });
+  }
+  public batch(contractInstance: Contract, calls: Call[]) {
+    calls.forEach((call: Call) => {
+      this.add(contractInstance, call);
+    });
+  }
+  // these requests only read contract state
+  public async read(_requests: Request[] = this.requests) {
+    const encodedRequests = _requests.map((request) => this.encodeRequest(request));
+    const { returnData } = await this.multicallClient.callStatic.aggregate(encodedRequests);
+    const zipped = zip(_requests, returnData);
+    return zipped.map(([request, response]) => {
+      if (request && response) return this.decodeResponse(request, response);
+      throw new Error("Unable to decode contract response");
+    });
+  }
+}
+
+// Multicall class that exposes public functions to the user and recursively chains itself.  Acts immutable
+// if you store reference to the parent intsance. Children will contain mutated state.
+export class Multicall {
```

Comment:
This implementation is very cool!

The big question I have is why we need the Core class since it seems to be transient? It seems like we could just dump all of that functionality into this class, right?

One suspicion I have for why you might've done it this way is because you want to expose information to the other Multicall instance that you don't want to expose to the user. Note: you can use private in a fairly clever way to achieve this without the other class.

Let's say you have two instances of Multicall. Each one has a private method `foo()` and private data `bar`. The following setup is totally allowed because different instances of the same class can access each others' private information:
```ts
export class Multicall {
  private number: bar;
  
  private foo() {
    console.log("called a private method!");
  }

  public passOtherMultiCall(otherMulticall: Multicall) {
    otherMultiCall.foo();
    console.log("some private data", othermulticall.bar);
  }
}
```

See https://www.typescriptlang.org/docs/handbook/2/classes.html#cross-instance-private-access for more info.

I think this means you could wrap all the mutation functionality into a private method or two within this class and call it from the other instance when needed. Thoughts?